referenceframe/urdf.go

```diff
@@ -0,0 +1,258 @@
+// between the two kinematics encoding schemes.
+func ConvertURDFToConfig(xmlData []byte, modelName string) (Model, error) {
+	// empty data probably means that the read URDF has no actionable information
+	if len(xmlData) == 0 {
+		return nil, ErrNoModelInformation
+	}
+
+	urdf := &URDFConfig{}
+	err := xml.Unmarshal(xmlData, urdf)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to convert URDF data to equivalent URDFConfig struct")
+	}
+
+	// Code below this point could be split off into another function, similarly to ParseConfig
+	if modelName == "" {
+		modelName = urdf.Name
+	}
+
+	model := NewSimpleModel(modelName)
+	transforms := map[string]Frame{}
+
+	zeroPosition := r3.Vector{X: 0.0, Y: 0.0, Z: 0.0}
+	zeroOrientation := spatial.NewZeroOrientation()
+	zeroPose := spatial.NewPoseFromOrientation(zeroPosition, zeroOrientation)
+
+	// Handle joints
+	var firstLink string
+	childMap := map[string]string{}
+	parentMap := map[string]string{}
+	jointOrigins := map[string]spatial.Pose{}
+
+	for _, jointElem := range urdf.Joints {
+		jointName := jointElem.Name
+
+		// TODO(wspies): Consider this joint more carefully
+		if strings.Contains(jointName, "world") {
+			firstLink = jointElem.Child.Link
+			continue
+		}
+
+		// Parse important details about each joint, including axes, limits, and relationships
+		jointAxes := convStringAttrToFloats(jointElem.Axis.XYZ)
+		jointLimits := Limit{Min: jointElem.Limit.Lower, Max: jointElem.Limit.Upper}
+		childMap[jointElem.Parent.Link] = jointName  // This will get used to find the terminal link(s) quickly
+		parentMap[jointElem.Child.Link] = jointName
+		parentMap[jointName] = jointElem.Parent.Link
+
+		// Generate transform data for the given joint type
+		switch jointElem.Type {
+		case "continuous", "revolute":
+			if jointElem.Type == "continuous" {
+				jointLimits.Min, jointLimits.Max = -math.Pi, math.Pi  // TODO(wspies): PR comment bait for limit constants
+			}
+			transforms[jointName], err = NewRotationalFrame(jointName,
+				spatial.R4AA{RX: jointAxes[0], RY: jointAxes[1], RZ: jointAxes[2]}, jointLimits)
+		case "prismatic":
+			transforms[jointName], err = NewTranslationalFrame(jointName,
+				r3.Vector{X: jointAxes[0], Y: jointAxes[1], Z: jointAxes[2]}, jointLimits)
+		case "fixed":
+			transforms[jointName], err = NewStaticFrame(jointName, zeroPose)
+		default:
+			return nil, errors.Errorf("Unsupported joint type detected: %v", jointElem.Type)
+		}
+
+		// Preserve transform data for a given joint so the parent link can create a static frame more easily later, since
+		// URDFs manage frame origins within joints instead of within links
+		geoXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+		geoRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+		jointPose := spatial.NewPoseFromOrientation(r3.Vector{X: geoXYZ[0], Y: geoXYZ[1], Z: geoXYZ[2]},
+			&spatial.EulerAngles{Roll: geoRPY[0], Pitch: geoRPY[1], Yaw: geoRPY[2]})
+		jointOrigins[jointName] = jointPose
```

Comment:
I dont think you should need all this.  I might not be understanding something though.  Will continue to read